### PR TITLE
WIP: Integrate replay logging into normal operation, backend for replay playback, and logged test data with Referee messages

### LINKED ISSUE #1625: Integrate replay logging
### Acceptance criteria

- [ ] SensorMsgs are logged during normal operation
- [ ] Dynamic parameters to select whether to log, and what directory to log to

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

-Integrate replay logging into `full_system_main` with dynamic parameters to control its operation
-Replay backend with a command line argument for the replay directory that populates a dynamic parameter
-added a `time_received` timestamp as a member of `SensorMsg` to enable real-time playback timing
-Updated logged data used for unit testing

### Testing Done

- Works on my machine

### Resolved Issues

Resolves #1625, #1624, #1584, #1452 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/backend/replay_backend.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        auto this_msg_received_time = std::chrono::duration<double>(
+            sensor_msg_or_null->time_received().epoch_timestamp_seconds());
+
+        if (last_msg_received_time && last_msg_replayed_time)
+        {
+            std::this_thread::sleep_until(
+                *last_msg_replayed_time +
+                (this_msg_received_time - *last_msg_received_time));
+        }
+        this->sendValueToObservers(*sensor_msg_or_null);
```
Comment (jonathanlew) [inline src/software/backend/replay_backend.cpp:38]:
```suggestion
        Subject<SensorMsg>::sendValueToObservers(*sensor_msg_or_null);
```

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Did a quick pass, let us know when you want a full review

File: src/software/backend/replay_backend.cpp (hunk 1 above)
Comment (jordanzliu) [reply src/software/backend/replay_backend.cpp:38]: I think SFINAE will result in the correct function being called?

Comment (jonathanlew) [reply src/software/backend/replay_backend.cpp:38]: yeah I'm sure it's correct, but it's more clear to invoke the superclass. Not a blocking comment

**Review by jonathanlew — COMMENTED** [review]
Left an idea for how to have more type checking, lmk what you think
Please merge in upstream/master since it cleans up the generic factory a bit with `TYPENAME` replacing ::name

**Comment by jordanzliu** [comment]
Will redo this PR after #1723 is merged

**Pull request closed without merging**